File: climbing-stairs/eunhwa99.java
// 시간 복잡도 O(n) - n은 계단의 개수
// 공간 복잡도 O(1) - 상수 공간 사용
class Solution{
    public int climbStairs(int n){
        // 피보나치
        int prev1 = 1; // 0번째 계단
        int prev2 = 1; // 1번째 계단

        for(int i=2; i<=n; i++){
            int current = prev1 + prev2;
            prev1 = prev2;
            prev2 = current;
        }
        return prev2;
    }
}

File: insert-interval/eunhwa99.kt
class Solution {
    fun insert(intervals: Array<IntArray>, newInterval: IntArray): Array<IntArray> {
        val result = mutableListOf<IntArray>()
        var currentStart = newInterval[0]
        var currentEnd = newInterval[1]
        var isAdded = false
        
        // TC: O(n)
        // SC: O(n)
        for(interval in intervals){
            if(isAdded) {
                result.add(interval)
            } else {
                if(interval[1] < currentStart){ // new interval is after the current interval
                    result.add(interval) // just add the interval without any changes
                }
                else if(interval[0] > currentEnd){ // new interval is before the current interval
                    result.add(intArrayOf(currentStart, currentEnd))
                    result.add(interval)
                    isAdded = true
                }
                else{ // intervals overlap, merge them
                    currentStart = minOf(interval[0], currentStart)
                    currentEnd = maxOf(interval[1], currentEnd)
                }
            }
        }
        
        // If new interval wasn't added yet, add it now
        if(!isAdded) {
            result.add(intArrayOf(currentStart, currentEnd))
        }
        
        return result.toTypedArray()
    }
}
